### ISSUE TITLE
operator: nil pointer access in reconcileDeploymentChanges

### ISSUE THREAD

**Issue by pohly** [intel/pmem-csi#617]
During testing, I noticed this error output for the operator pod:
```
pmem-csi-operator-6747bf59fd-jgzxk/pmem-csi-operator@pmem..ker2: k8s.io/apimachinery/pkg/util/runtime.logPanic(0x152d660, 0x23bc3c0)
pmem-csi-operator-6747bf59fd-jgzxk/pmem-csi-operator@pmem..ker2: 	/src/pmem-csi/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:74 +0xa3
pmem-csi-operator-6747bf59fd-jgzxk/pmem-csi-operator@pmem..ker2: k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
pmem-csi-operator-6747bf59fd-jgzxk/pmem-csi-operator@pmem..ker2: 	/src/pmem-csi/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:48 +0x82
pmem-csi-operator-6747bf59fd-jgzxk/pmem-csi-operator@pmem..ker2: panic(0x152d660, 0x23bc3c0)
pmem-csi-operator-6747bf59fd-jgzxk/pmem-csi-operator@pmem..ker2: 	/go/src/runtime/panic.go:679 +0x1b2
pmem-csi-operator-6747bf59fd-jgzxk/pmem-csi-operator@pmem..ker2: github.com/intel/pmem-csi/pkg/pmem-csi-operator/controller/deployment.(*PmemCSIDriver).reconcileDeploymentChanges(0xc0001d7c00, 0xc00021e190, 0x0, 0xc0002a24e0, 0xc0001d7a00, 0x2, 0x2, 0x16fc333)
pmem-csi-operator-6747bf59fd-jgzxk/pmem-csi-operator@pmem..ker2: 	/src/pmem-csi/pkg/pmem-csi-operator/controller/deployment/controller_driver.go:151 +0x17e
pmem-csi-operator-6747bf59fd-jgzxk/pmem-csi-operator@pmem..ker2: github.com/intel/pmem-csi/pkg/pmem-csi-operator/controller/deployment.(*PmemCSIDriver).Reconcile(0xc0001d7c00, 0xc00021e190, 0x27, 0x0, 0x0)
pmem-csi-operator-6747bf59fd-jgzxk/pmem-csi-operator@pmem..ker2: 	/src/pmem-csi/pkg/pmem-csi-operator/controller/deployment/controller_driver.go:94 +0x43a
pmem-csi-operator-6747bf59fd-jgzxk/pmem-csi-operator@pmem..ker2: github.com/intel/pmem-csi/pkg/pmem-csi-operator/controller/deployment.(*ReconcileDeployment).Reconcile(0xc00021e190, 0x0, 0x0, 0xc0000475d8, 0x8, 0xc00002ec00, 0x0, 0x0, 0x0)
pmem-csi-operator-6747bf59fd-jgzxk/pmem-csi-operator@pmem..ker2: 	/src/pmem-csi/pkg/pmem-csi-operator/controller/deployment/deployment_controller.go:167 +0x288
pmem-csi-operator-6747bf59fd-jgzxk/pmem-csi-operator@pmem..ker2: sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc0001bc000, 0x15851a0, 0xc00000e480, 0x0)
pmem-csi-operator-6747bf59fd-jgzxk/pmem-csi-operator@pmem..ker2: 	/src/pmem-csi/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:256 +0x162
pmem-csi-operator-6747bf59fd-jgzxk/pmem-csi-operator@pmem..ker2: sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc0001bc000, 0x411300)
pmem-csi-operator-6747bf59fd-jgzxk/pmem-csi-operator@pmem..ker2: 	/src/pmem-csi/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:232 +0xcb
pmem-csi-operator-6747bf59fd-jgzxk/pmem-csi-operator@pmem..ker2: sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker(0xc0001bc000)
pmem-csi-operator-6747bf59fd-jgzxk/pmem-csi-operator@pmem..ker2: 	/src/pmem-csi/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:211 +0x2b
pmem-csi-operator-6747bf59fd-jgzxk/pmem-csi-operator@pmem..ker2: k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xc00003c570)
pmem-csi-operator-6747bf59fd-jgzxk/pmem-csi-operator@pmem..ker2: 	/src/pmem-csi/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:155 +0x5e
pmem-csi-operator-6747bf59fd-jgzxk/pmem-csi-operator@pmem..ker2: k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc00003c570, 0x18f0c60, 0xc0002a2390, 0xc000010201, 0xc0001e4780)
pmem-csi-operator-6747bf59fd-jgzxk/pmem-csi-operator@pmem..ker2: 	/src/pmem-csi/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:156 +0xa3
pmem-csi-operator-6747bf59fd-jgzxk/pmem-csi-operator@pmem..ker2: k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc00003c570, 0x3b9aca00, 0x0, 0x1, 0xc0001e4780)
pmem-csi-operator-6747bf59fd-jgzxk/pmem-csi-operator@pmem..ker2: 	/src/pmem-csi/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0xe2
pmem-csi-operator-6747bf59fd-jgzxk/pmem-csi-operator@pmem..ker2: k8s.io/apimachinery/pkg/util/wait.Until(0xc00003c570, 0x3b9aca00, 0xc0001e4780)
pmem-csi-operator-6747bf59fd-jgzxk/pmem-csi-operator@pmem..ker2: 	/src/pmem-csi/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:90 +0x4d
pmem-csi-operator-6747bf59fd-jgzxk/pmem-csi-operator@pmem..ker2: created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1
pmem-csi-operator-6747bf59fd-jgzxk/pmem-csi-operator@pmem..ker2: 	/src/pmem-csi/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:193 +0x328
pmem-csi-operator-6747bf59fd-jgzxk/pmem-csi-operator@pmem..ker2: panic: runtime error: invalid memory address or nil pointer dereference [recovered]
pmem-csi-operator-6747bf59fd-jgzxk/pmem-csi-operator@pmem..ker2: 	panic: runtime error: invalid memory address or nil pointer dereference
pmem-csi-operator-6747bf59fd-jgzxk/pmem-csi-operator@pmem..ker2: [signal SIGSEGV: segmentation violation code=0x1 addr=0x180 pc=0x1389f9e]
```

I don't know exactly what's triggering it, and it doesn't seem to break any existing E2E test. The code is:

https://github.com/intel/pmem-csi/blob/4ea70c5532f8bbb61e5a17f3eb0bd83744f8ec89/pkg/pmem-csi-operator/controller/deployment/controller_driver.go#L61-L92

I'm not sure how exactly the code is supposed to work. If `r.deployments` does not contain the name, `oldDeployment = nil, foundInCache = false`. That then directly gets passed to `d.reconcileDeploymentChanges` which uses `existing` without checking for `nil`:
https://github.com/intel/pmem-csi/blob/4ea70c5532f8bbb61e5a17f3eb0bd83744f8ec89/pkg/pmem-csi-operator/controller/deployment/controller_driver.go#L149

**Comment by pohly** [comment]
We don't seem to have a test for this. The new test from https://github.com/intel/pmem-csi/pull/610, "should be able to capture deployment changes when operator is not running", only covers changes that are allowed.

We need a test that resets invalid changes to invariant spec members (and only those, see issue #615) that were made while the operator was down.

**Comment by pohly** [comment]
We also need a test for "new deployment added while operator was down". There is code for that (https://github.com/intel/pmem-csi/blob/4ea70c5532f8bbb61e5a17f3eb0bd83744f8ec89/pkg/pmem-csi-operator/controller/deployment/controller_driver.go#L1132-L1134), it just isn't getting tested.

In order to be able to create such a deployment, we probably need to:
- create one deployment
- stop the operator
- create another one
- restart the operator
because otherwise the CRD is not installed when we try to test.

**Comment by pohly** [comment]
Resolved in PR #623
